Audio and playlist components

chmorgan opened the pull request:
Refactor audio player and playlist out of mp3_demo and make use of both in mp3_demo and factory examples.


@TDA-2030 I'd like to get the audio_player up on the components catalog at some point as I could make use of this on another project. Its not ready quite yet and lacks WAV file support (which I can add soon) but wanted to get your initial review and feedback to start moving in that direction.

chmorgan committed:
mp3_demo - Remove unused @brief entries and remove specific files from @file entries as doxygen uses the filename if the filename is not specified

chmorgan committed:
playlist component

chmorgan committed:
audio_player component

chmorgan committed:
mp3_demo - ui_audio - Update for playlist and audio component changes, track playback state on song changes

chmorgan committed:
mp3_demo - Changes for audio subsystem and playlist components

chmorgan committed:
factory example - Switch to using audio_player and playlist components

chmorgan commented:
I've got some fixes and cleanups that I missed and I'm adding wav support now. I'll msg here when its pushed and ready for review.

chmorgan commented:
@TDA-2030 any reason that esp-adf wasn't used initially? I was unaware that esp-adf was so developed when I started on this a few weeks ago. Looking at it looks like the player and playlist stuff could be swapped out with esp-adf. One downside is that esp-adf doesn't provide codec source code, which is odd no?. It was slightly disappointing that I wasn't aware of esp-idf until now and perhaps a little work was wasted although it was fun none the less.

Otherwise I was tempted to complete what I've got, which is mp3 + wav file support, some fixes to play the end of the file (today we actually trim the end of the file off), and slightly better buffer support for mp3 playback so we aren't fread() + memmove() on each loop.

Thoughts? Should I switch to integrate esp-adf instead? I was hoping to get your feedback.

TDA-2030 commented:
`esp-box` is not primarily used for audio applications, and `esp-adf` is not a lightweight library, so it is not integrated into the `esp-box`. If you use a lot of audio-related functionality, integrating `esp-adf` into your project will really help you get things done faster and better.

chmorgan commented:
@TDA-2030 makes sense, appreciate the info. I'm getting close to having the audio_component ready. It's simple and fully open source, so it should have some utility for some people that are looking to avoid perhaps the overhead of esp-adf and I'm going to at least get the pr ready here before digging into trying out esp-adf.

chmorgan commented:
@TDA-2030 alright, here is the next rev and I think ready for review. For my other use case I'm going to integrate another audio codec, a TI one, and that may drive some improvements to the audio component, making it more generic or whatever. But this is I think a good point to review.

TDA-2030 commented:
It seems to be a great improvement. Thank you for your contribution!

TDA-2030 commented:
> For my other use case I'm going to integrate another audio codec, a TI one

Can you share some information about the audio codec you mentioned? Is it open source and free?

TDA-2030 commented on components/audio_player/audio_player.cpp:
```diff
@@ -0,0 +1,875 @@
+/**
```
Why .cpp instead .c？

chmorgan replied:
So the symbols and functions are in c context, if you look at the header file. I picked .cpp for the source file because I wanted the compiler to better enforce type safety, basically trying to provide flexibility to people with c and c++ codebases to use the component while also adding type safety within that file.

Does using .cpp mess up some projects that aren't enabling cpp usage? Are there things I should and can do to reduce the unintended impact of using the c++ compiler there?

chmorgan commented:
> > For my other use case I'm going to integrate another audio codec, a TI one
> 
> Can you share some information about the audio codec you mentioned? Is it open source and free?

Oh I mean hardware codec. It's a TI codec but I don't recall the part number at the moment. We tried to source the evergreen es8311 but had trouble finding a US source for that component. If you've got one, or know of a source for that component that can converse in English I think we'd consider using the 8311.

There isn't any way this audio component can and should aim to really overlap with the esp-adf, that codebase is massively more developed. I guess I could see someone interested in ogg decoding etc but yeah if you are going with BT streaming etc you might as well use esp-adf :-)

chmorgan commented:
@TDA-2030 I should probably list out some of the improvements in the audio side.

- mp3 playback finishes the song now, previous implementation I think would stop and not play the last few frames when EOF was hit.
- mp3 playback reads in chunks, filling the buffer and then decoding from the buffer. Results in overall fewer filesystem reads although from a predictability standpoint it does mean cpu usage is slightly bursty due to the period reads to fill the buffer. It seemed worth the tradeoff vs. reading each loop.
- wave decoding, I needed this for some audio files from google's test to speech system for audio prompts
- maintains stereo / mono conversion, I think this should be an input parameter at some point as to whether you need the output in mono format
- still trying to address a slight tick when audio is changing songs. Attempt was through the mute function so you can turn off the codec when no audio is playing.
- Cleaned up events to better indicate state transitions, in particular playing -> playing without stopping. There is an ascii diagram at the top of audio_player.h to help the reader.
- Tried to clean up the language around 'samples' vs. 'frames', I think its consistent and mostly correct at this point.
- Refactored the playlist management out. Then noticed esp-adf has basically identical code, I guess great minds think alike? :)
- Added layers of logging, this approach was borrowed maybe from the BT component? Where each log level produces more verbose output for debugging, configurable via kconfig.
- Moved allocation to audio_player_init(), if you don't call this a number of buffers aren't allocated. We could statically allocate some of these, or ask the user to pass them in if the user would like to decide where/how to allocate this memory. Thoughts? There is presently no deinit function and no mechanism to have more than one audio component instance although this would be possible if we wrapped things up into an instance struct that the caller had to pass in for each call. Thoughts?

I think this pretty much summarizes it. Sorry for the delay. I also appreciate it overlaps with some esp-adf stuff but like your earlier comment the project I'm working on has no need for the full esp-adf framework and its hard to know how the esp-adf internals work or how much overhead there is.

One thing could be to look at the api there more closely to see if this code can be adjusted to better match it, or if its just silly duplicating this functionality here and if esp-adf can be used in a lightweight manner. It was reasonably fun to write this stuff and its scope is limited. Anyway, if we don't need two of the same thing then why have it? But it sounds like this is a lot more compact so....

Looking forward to your other thoughts and what you'd like to see to get it ready for merge.

chmorgan commented:
Hi @TDA-2030 any chance to take a look at this PR? I was hoping to get it moving along the review and merging process.

chmorgan commented:
Ping on review and feedback.

ESP-Mars commented:
> Hi @TDA-2030 any chance to take a look at this PR? I was hoping to get it moving along the review and merging process.

> Ping on review and feedback.

Sorry, TDA-2030 has left Espressif, will arrange another guy to tackle the problems.

chmorgan commented:
@ESP-Mars I'd be willing to be added as a maintainer to this repo. We do need another reviewer for this PR in any case. I have at least one other cleanup PR I'll be pushing up in the near future as well.

tore-espressif commented on components/audio_player/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+idf_component_register(SRCS "audio_player.cpp"
+                       INCLUDE_DIRS "include"
+                       REQUIRES bsp libhelix-mp3
```
Could we remove the dependency on `bsp` ? If yes, the `audio_player` could be a nice reuseble component

tore-espressif commented on components/audio_player/Kconfig:
```diff
@@ -0,0 +1,20 @@
+        bool "Enable wav file playback"
+        default y
+        help
+            Wave playback coming soon!
```
Isn't this done already?

tore-espressif commented on components/audio_player/include/audio_player.h:
```diff
@@ -0,0 +1,150 @@
+#include <stddef.h>
+#include <stdio.h>
+#include "esp_err.h"
+#include "driver/i2s.h"
```
Just an FYI:
`driver/i2s.h` driver will become obsolete in IDF version 5.0. You can still use it, but you'll get deprecation warnings (they can be turned off). ATM, I'd stick to the old driver, because it works on all active esp-idf releases.

More information is in our [migration guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/migration-guides/release-5.x/peripherals.html#i2s-driver)

tore-espressif reviewed (commented):
Hello @chmorgan , thank you very much for you patience.

> If you've got one, or know of a source for that component that can converse in English I think we'd consider using the 8311.

If you are still interested in ES8311, I'm attaching full documentation that we got from Everst Semi (it contains the 'Confidential' watermark, but we have a written permission to share this with users). So far, I haven't found any 'documentation holes'. We also have a simple [driver for ES8311](https://github.com/espressif/esp-bsp/tree/master/components/es8311), (it gets uploaded to https://components.espressif.com/component/espressif/es8311 we'd like to reuse this driver in all esp implementations) 

[ES8311 user Guide.pdf](https://github.com/espressif/esp-box/files/9314453/ES8311.user.Guide.pdf)
[ES8311 DS.pdf](https://github.com/espressif/esp-box/files/9314454/ES8311.DS.pdf)

When it comes to the PR itself, I really like how you created separate components that can be reused in other projects too!

> We could statically allocate some of these, or ask the user to pass them in if the user would like to decide where/how to allocate this memory. Thoughts?

Since this component is intended for use with esp-idf, and we use memory allocation in esp-idf a lot, I'd stick with dynamic allocation. I guess that the memory must be DMA capable, right? That also means that there should be a deinit function that cleans-up after the component. (Nitpick: If you want to be aligned with esp-idf nomenclature, it should be _new / _delete instead of _init / _deinit -> it implies dynamically allocated object). 

> no mechanism to have more than one audio component instance

Is there any usecase where user might need two instances? We can define this object 'singleton by design' and/or add support for multiple instances later (that would be a breaking change)

chmorgan commented:
@tore-espressif updated PR coming in the next few days. Lots of cleanups and improvements, working on getting tests now although I ended up symlinking into the esp-idf/ install folder, couldn't figure out a better way to get unit-test-app to work correctly.

For this to be a component should I push it to a repo on my account or is that something we discuss later?

tore-espressif commented:
> updated PR coming in the next few day

Looking forward to it!

> couldn't figure out a better way to get unit-test-app to work correctly.

Yeah, we are just finishing improvements that would make this easier for external users

> For this to be a component should I push it to a repo on my account

Since you are the author and maintainer, I'd suggest keeping this under your namespace (and license) on GitHub. Then we can grant you write access to our [components registry](https://components.espressif.com/)  that would allow you to push (and then reuse) these components in any esp-idf projects. Does that sound reasonable?

chmorgan commented:
@tore-espressif alright, pushed an updated branch with a ton of changes, plus unity tests, and a README.md. Next step after your next round of review is to move this out into its own repo and pull it in as a component, but thought we should continue review here until its ready. If you could take another look I'd appreciate it!

tore-espressif reviewed (commented):
Hello @chmorgan very nice work, I like the clean dependencies and APIs, just the necessary #includes and C++ guard macros.

I left few comments, but there is no blocker. Some of them are just my opinions/questions so feel free to disregard them. 

Anyway, I'd like to try building a simple application using these components on my own, I might see something that I missed now...

> Next step after your next round of review is to move this out into its own repo and pull it in as a component, but thought we should continue review here until its ready.

yep, I agree. Thank you again! looking forward to hearing from you

chmorgan commented on components/audio_player/include/audio_player.h:
```diff
@@ -0,0 +1,172 @@
+    AUDIO_PLAYER_UNMUTE
+} AUDIO_PLAYER_MUTE_SETTING;
+
+typedef esp_err_t (*audio_player_mute_fn)(AUDIO_PLAYER_MUTE_SETTING setting);
```
Generally I always advocate enums in place of booleans, otherwise its unclear whether true is muted or not muted etc.

chmorgan commented on components/audio_player/test/audio_player_test.c:
```diff
@@ -0,0 +1,191 @@
+TEST_CASE("audio player can be newed and deleted", "[audio player]")
+{
+    esp_err_t ret = audio_player_new(I2S_NUM_0, audio_mute_function);
+    TEST_ASSERT_TRUE(ret == ESP_OK);
```
doh, I didn't realize that existed but should have checked. Not sure how I didn't catch that....

chmorgan commented on components/audio_player/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+idf_component_register(SRCS "audio_player.cpp" "audio_mp3.cpp" "audio_wav.cpp"
+                       INCLUDE_DIRS "include"
+                       REQUIRES libhelix-mp3
```
So, If I adjust this like the unity example I get compilation errors, unable to find a libhelix-mp3 include file, mp3dec.h. However, MESSAGE() indicates that the if(CONFIG_AUDIO_PLAYER_ENABLE_MP3) is being visited, the MESSAGE() is printed out.

It's almost like there is some kind of odd dependency issue where an initial pass is generating some makefiles and a second pass is finding them. I can see this in the cmake output from idf.py build.

And libhelix-mp3 is not included in the include search paths when the compilation fails (which explains why it can't find mp3dec.h).

Using v4.4.2 here, maybe this has been fixed on master?


Oh, and oddly enough it works correctly with adding the audio_wav.cpp and audio_mp3.cpp files, that's not an issue, those files are found and built. So it does look like some odd multi pass thing where the CONFIG options aren't present in the first pass (which causes libhelix-mp3 to not be a dependency, which means its include paths aren't added), but are present on the second pass.

I added some MESSAGE() lines, one inside of the CONFIG_ section and one before idf_component_register()

```
cmorgan@Chriss-MBP mp3_demo % idf.py reconfigure
Executing action: reconfigure
Running cmake in directory /Users/cmorgan/projects/esp-box/examples/mp3_demo/build
Executing "cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -DIDF_TARGET=esp32s3 -DCCACHE_ENABLE=0 /Users/cmorgan/projects/esp-box/examples/mp3_demo"...
-- Found Git: /usr/local/bin/git (found version "2.37.1") 
-- Component directory /Users/cmorgan/projects/esp-box/components/esp-rainmaker does not contain a CMakeLists.txt file. No component will be added
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /Users/cmorgan/.espressif/tools/xtensa-esp32s3-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s3-elf/bin/xtensa-esp32s3-elf-gcc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Users/cmorgan/.espressif/tools/xtensa-esp32s3-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s3-elf/bin/xtensa-esp32s3-elf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Users/cmorgan/.espressif/tools/xtensa-esp32s3-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32s3-elf/bin/xtensa-esp32s3-elf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building ESP-IDF components for target esp32s3
-- *************** requires is        <---------------- (missing here, but see below where its set)
-- Project sdkconfig file /Users/cmorgan/projects/esp-box/examples/mp3_demo/sdkconfig
Loading defaults file /Users/cmorgan/projects/esp-box/examples/mp3_demo/sdkconfig.defaults...
-- Found Python3: /usr/local/Frameworks/Python.framework/Versions/3.9/bin/python3.9 (found version "3.9.13") found components: Interpreter 
-- Performing Test C_COMPILER_SUPPORTS_WFORMAT_SIGNEDNESS
-- Performing Test C_COMPILER_SUPPORTS_WFORMAT_SIGNEDNESS - Success
-- App "mp3_demo" version: v0.3.0-18-g0ca9c80-dirty
-- Adding linker script /Users/cmorgan/projects/esp-box/examples/mp3_demo/build/esp-idf/esp_system/ld/memory.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_system/ld/esp32s3/sections.ld.in
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/esp_rom/esp32s3/ld/esp32s3.rom.newlib-time.ld
-- Adding linker script /Users/cmorgan/esp/esp-idf/components/soc/esp32s3/ld/esp32s3.peripherals.ld
-- IPV6 support is disabled so the asio component will not be built
-- *********************mp3 support enabled           <--------------------------
-- *************** requires is libhelix-mp3                 <----------------
```

chmorgan replied:
@tore-espressif I just put a todo comment in the CMakeLists.txt and sticking with that for the time being. When/if its fixed I'll move it back into the if(). I don't understand the build system well yet.

chmorgan commented on components/audio_player/audio_player.cpp:
```diff
@@ -0,0 +1,595 @@
+                                "Audio Task",
+                                4 * 1024,
+                                &instance,
+        (UBaseType_t)           configMAX_PRIORITIES - 1,
```
Never hurts to have all of the cpu, and starve everything else!  Just kidding, added it to a config struct.

chmorgan commented:
@tore-espressif very good review feedback, I appreciate it a ton. I should have an updated PR coming soon but also plan to split things into separate repos at this point, and also noticed libhelix-mp3 has a ton of duplicated files and needs to be its own component repo with a submodule, which I'll put in place as a part of the next round as well.
